### PR TITLE
Hyper-parameter optimization 2.0

### LINKED ISSUE #20: hyperparameter optimization 2.0
Implement hyper paramter folder structure.

It might be feasible to precompute a lot of reservoirs with the following parameters:
    * hdim: 5k, 10k, 15k, 20k
    * precompute the spectral radius to scale later
    * density: 1%
    * input weight init

### PR DESCRIPTION
**Pull request by nmheim**
Trying to resolve #20 

The hyper-parameter optimization will create a folder structure parameter. If the hyper-params are `spectral_radius`, and `weight_init` it will look like this:

```
model-spectral_radius:1.5-weight_init:1.0/
     - params.json
     - model.pth
     - train_data.nc
     - run-domain:{domain}-transient_length:{transient_length}/
     - ...
```

After the __untrained__ `model.pth` has been created all following steps are deterministic as the only source of randomness is the creation of the model matrices.
The `train_data.nc` of the run folders will contain the used inputs/labels and created states which can now be generated.

Next the output layer can be optimized. For some (level-two) hyper-parameters the optimization of the states that are created by a given input sequence are invariant and can be reused. This is reflected in another sub directory of the folder structure:

```
run-transient_length:{transient_length}/
    - pinv/
    - tik{tikhonov_beta}
```

These subdirectories will contain `pinv` and `tik` folders. They contain the now __trained__ models and the corresponding predictions. Additionally, the `params.json` in which only the parameters defined in the parent paths are updated accordingly.
```
pinv/
    - pred_data.nc
    - model.pth
    - params.json
```

### CODE REVIEW

**Comment by nmheim** [comment]
what do you think @jamesavery ?
